src/server_manager/electron_app/digitalocean_oauth.ts

```diff
@@ -0,0 +1,218 @@
+      .toString('hex')  // convert to hexadecimal format
+      .slice(0, len);   // return required number of characters
+}
+
+interface ServerError extends Error {
+  code: string;
+}
+
+// Makes server listen on each of the listed ports until there's one open.
+// Returns the index of the port used.
+function listenOnFirstPort(server: http.Server, portList: number[]): Promise<number> {
+  let portIdx = 0;
+  return new Promise((resolve, reject) => {
+    server.once('listening', () => {
+      console.log(`Listening on port ${portList[portIdx]}`);
+      resolve(portIdx);
+    });
+    server.on('error', (error: ServerError) => {
+      if (error.code === 'EADDRINUSE') {
+        console.log(`Port ${portList[portIdx]} already in use`);
+        portIdx += 1;
+        if (portIdx < portList.length) {
+          const port = portList[portIdx];
+          console.log(`Trying port ${port}`);
+          server.listen({host: 'localhost', port, exclusive: true});
+          return;
+        }
+      }
+      server.close();
+      reject(error);
+    });
+    server.listen({host: 'localhost', port: portList[portIdx], exclusive: true});
+  });
+}
+
+interface Account {
+  droplet_limit: number;
+  floating_ip_limit: number;
+  email: string;
+  uuid: string;
+  email_verified: boolean;
+  status: string;
+  status_message: string;
+}
+
+function getAccount(accessToken: string): Promise<Account> {
+  return new Promise((resolve, reject) => {
+    request(
+        {
+          url: 'https://api.digitalocean.com/v2/account',
+          headers: {
+            'User-Agent': 'Outline Manager',
+            'Content-Type': 'application/json',
+            'Authorization': `Bearer ${accessToken}`
+          }
+        },
+        (error, response, body) => {
+          if (error) {
+            return reject(error);
+          }
+          const bodyJson: {account: Account} = JSON.parse(body);
+          return resolve(bodyJson.account as Account);
+        });
+  });
+}
+
+export interface OauthSession {
+  // The result of the OAuth session, with the authentication token.
+  result: Promise<string>;
+  // Checks if the OauthSession has been cancelled.
+  isCancelled(): boolean;
+  // Cancels the OAuth session. The result promise is automatically rejected
+  // and isCancelled() will return true.
+  cancel(): void;
+}
+
+// Runs the DigitalOcean oauth flow and returns the access token.
+// See https://developers.digitalocean.com/documentation/oauth/ for the API.
+export function runOauth(): OauthSession {
+  const secret = randomValueHex(16);
+
+  const app = express();
+  const server = http.createServer(app);
+  server.on('close', () => console.log('Oauth server closed'));
+
+  let isCancelled = false;
+  app.use((req, res, next) => {
+    if (isCancelled) {
+      res.status(503).send('Authentication cancelled');
+    } else {
+      next();
+    }
+  });
+
+  app.get('/', (request, response) => {
+    response.send(`<html>
```

Comment:
This page is running in the user's browser, right?  Are they redirected to it?  Do they see a blank page? Are they left at the blank page until they take further action?

If the user can see this page, then we should probably try to give it some content.  Alternatively, we could have the page try to close itself when it's done.

Either way, that would add enough code that this html should probably be moved to a separate file.

Reply:
In practice this page is never really seen. It's immediately redirected. Because of that I think it's ok to not have any content.